Fix/merge 3187 319 run e2e

marcambier opened the pull request:
:warning: this is the raw 3.18.7 -> master merge :warning: 
:warning: it needs additionnal development in order to make the plan workflow works, related to https://github.com/gravitee-io/issues/issues/8168 :warning: 

**Issue**

https://github.com/gravitee-io/issues/issues/8389
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/fix-merge-3187-319-run-e2e/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-hnilvccbth.chromatic.com)
<!-- Storybook placeholder end -->
<!-- E2E Coverage placeholder -->
---
### 🧪 End-to-End Coverage

| INSTRUCTIONS | BRANCHES |
| :----------: | :------: |
|   37% | 22%   |

A more detailed report has been uploaded to [circleci](https://output.circle-artifacts.com/output/job/11873146-9fb8-459c-94cc-678ad2ecbbcf/artifacts/0/gravitee-apim-e2e/jacoco/reports/index.html)
<!-- E2E Coverage placeholder end -->

Commit:
fix: fix plans selection workflow

Before this change, when dealing with multiple plans based on the same security type, the policy is applied prior to retrieve the associated subscription that needs to checked (same security type -> JWT and Oauth2 are both based on Authorization bearer token).
This is not optimal because, in some circumstances, it can select a subscription that is not linked to the plan that has been executed.

This change refactor the plan selection workflow by reordering the different steps.

For JWT and Oauth2 plans, subscription has to be searched with client_id, to see if a plan can handle the incoming request.

For Oauth2, client_id is retreived from token introspection, so we can't check subscription outside of policy chain execution, where the auth server is called.

For JWT, client_id is always present in incoming request token claims. So we can fetch the subscription prior of policy chain execution.

JWT are now processed prior to Oauth2 plans.
Having multiple Oauth2 plans on the same API will still require selection rules to apply to right one.

https://github.com/gravitee-io/issues/issues/7824

Okhelifi committed:
ci: change slack notification channel

gcusnieux committed:
fix(heartbeat): resolve repository exceptions management

gravitee-io/issues#8267

gaetanmaisse committed:
refactor: remove useless info icon in dictionary creation screen

This icon has never been linked to anything

gcusnieux committed:
chore: bump expression-language to 1.9.5

gravitee-io/issues#8217

gcusnieux committed:
chore: bump gravitee-policy-regex-threat-protection to 1.2.2

gravitee-io/issues#8272

gaetanmaisse committed:
fix: properly handle login for a user with no environment role

A user without any role on any environment isn't able to access APIM.
So we now display a proper message and avoid ending in an infinite redirection loop.

https://github.com/gravitee-io/issues/issues/6532

gaetanmaisse committed:
style: fix position of notification number by setting line-height properly

gaetanmaisse committed:
fix: fix navigation from Org settings to My account screen

Use the default current env if there is no environment id in the url.

https://github.com/gravitee-io/issues/issues/6848

graviteeio committed:
3.10.19

Okhelifi committed:
Merge tag '3.10.19' into issues/#8279-merge-31019-into-315x

Okhelifi committed:
refactor: lint project after merging 3.10.19 into 3.15.x

gravitee-io/issues#8279

Okhelifi committed:
Merge 3.10.19 into 3.15.x

gravitee-io/issues#8279

Okhelifi committed:
chore: add missing enterprise edition profile for local run

phiz71 committed:
fix: filter headers according to exclude/include Rules

gravitee-io/issues#8305

phiz71 committed:
ci: exclude all APIM artefacts from cache

Change made to avoid cache growing. Indeed, because of the restore cache job, for each new version, we will get the cache from an older version with older artefacts that are not needed.
The cache will continue growing but far slower.

phiz71 committed:
perf: improve pages search

 - Use explicit filter on API or Environment when removing group reference from pages
 - add some new indices to improve indexation in databases

phiz71 committed:
chore: add run configurations for intelliJ

phiz71 committed:
fix(websocket): use wrapped handler instead of parent's one

gravitee-io/issues#8228

phiz71 committed:
fix(repository): accessControl fields should never be null

- initialize accessControl list in mongo.
- JDBC was already ok

gravitee-io/issues#8323

gaetanmaisse committed:
fix: bump `gravitee-connector-http` to `1.1.11`

https://github.com/gravitee-io/issues/issues/8228
https://github.com/gravitee-io/issues/issues/8324

phiz71 committed:
fix: use the right state name

gravitee-io/issues#8328

phiz71 committed:
fix: use the right field name for labels

gravitee-io/issues#8327

graviteeio committed:
chore: prepare next version [skip ci]

phiz71 committed:
fix: query has a impact on chart-line with same field

gravitee-io/issues#8325

phiz71 committed:
fix: make sidenav items clickable links

gravitee-io/issues#8326

gcusnieux committed:
fix(console): improve the search for apps during the subscription process

gravitee-io/issues#8329

Commit:
test(e2e): limit random group.name to 64 characters due to JDBC field limitation

gaetanmaisse committed:
fix: refactor the code used to `prefix` all MongoDB collections

Using Java 17, `setAccessible(true)` is throwing an error, so instead of using this we now use SpEL in @Document annotation

https://github.com/gravitee-io/issues/issues/7918

gaetanmaisse committed:
test: improve configuration of MongoDB repository tests

Use graviteeTest.yml file instead of hardcoded and/or default values

gaetanmaisse committed:
chore: add IntelliJ run config for MongoDB repository tests

Commit:
fix(jdbc): rate limit at platform level can use null subscription

It makes JDBC rate limit behave like mongo.
When using rate limit at platform policy level ; rate limit is not related to a subscription, so subscription in database has to be nullable.

Rate limit works nice even with a null subscription,
Key is composed by API / rateType / resolvedPath and makes it relevant.

ThibaudAV committed:
fix(console): identity providers - correct confirmation message

should be displayed, e.g.: "Identity provider successfully activated/deactivated"

phiz71 committed:
fix(portal): don't apply filters on an empty APIs list

ThibaudAV committed:
fix(console): do not open the documentation when pendo is enable

graviteeio committed:
3.18.6

graviteeio committed:
chore: prepare next version [skip ci]

Commit:
fix: make API out of sync banner consistent with "</" data in policy config

JsonSchemaService configuration validation escapes the slash in "</" string, leading to "<\/".

When deserializing a flow from API event, it's correctly interpreted as "</", thanks to jackson annotations on Step.configuration and custom deserializer.
When reading a flow from database, it's not correctly interpreted, as the inline mapping in flowConverter doesn't benefit of jackson annotations tricks.

This fixes makes flowConverter use jackson for step conversion, in order to have the step.configuration correctly interpreted.

This way, the synchronizationService doesn't detect any irrelevant difference between deployed API, and API to deploy;
And the API out of sync banner doesn't appear when it's not relevant.

https://github.com/gravitee-io/issues/issues/8343

graviteeio committed:
3.15.14

graviteeio committed:
chore: prepare next version [skip ci]

Commit:
chore(release): remove snyk manual step notice, since it's automated

Commit:
Merge tag '3.15.14' into fix/merge_315_318-2-run-e2e

Commit:
fix: this fixes 3.15.14 merge

marcambier committed:
Merge pull request #2365 from gravitee-io/fix/merge_315_318-2-run-e2e

Merge tag '3.15.14' into 3.18.x

Commit:
test: fix flaky E2E tests

Mock policy test is flaky, cause sometimes it checks the gateway response before the API was properly redeployed.

That has already been fixed on master branch ;
So this commit reports improvements from the master branch to fix that in 3.18 :
- reported the whole gateway.ts, including the new expectedResponseValidator parameter
- reported mock-policy gateway assertions, using the expectedResponseValidator

a-cordier committed:
fix: discard empty groups when getting user membership

see https://github.com/gravitee-io/issues/issues/8344

Commit:
fix: provide properly instantiated subscriptionRepository to PlanBasedAuthenticationHandlerEnhancer

In 3.10, PlanBasedAuthenticationHandlerEnhancer was instantiated in a spring configuration file ;
So the SubscriptionRepository was instantiated properly, thanks to @Autowired.

Now, PlanBasedAuthenticationHandlerEnhancer is instantiated from ApiContextHandlerFactory,
So we can't use @Autowired nomore, and have to provide it's SubscriptionRepository from context.

Also removing the unused component FreePlanAuthenticationHandlerEnhancer.

https://github.com/gravitee-io/issues/issues/8364

graviteeio committed:
3.15.15

Commit:
Merge tag '3.15.15' into fix/merge_31515_318

marcambier committed:
Merge pull request #2373 from gravitee-io/fix/merge_31515_318

Merge tag '3.15.15' into fix/merge_31515_318

a-cordier committed:
chore: bump policy xslt version to 1.6.2

see https://github.com/gravitee-io/issues/issues/8382

Commit:
test: fix websocket flaky tests by dedicate a specific port to each

Commit:
test: fix websocket flaky tests by using proper VertxTestContext assertions

graviteeio committed:
3.18.7

Commit:
Merge tag '3.18.7' into fix/merge-3187-319-run-e2e

Commit:
fix: fix 3.18.7 merge

gaetanmaisse committed:
fix: bump `gravitee-connector-http` to `2.0.5`

Related to:
https://github.com/gravitee-io/issues/issues/8228
https://github.com/gravitee-io/issues/issues/8324

gaetanmaisse committed:
fix: bump `gravitee-connector-api` to `1.1.4`

phiz71 commented on gravitee-apim-gateway/gravitee-apim-gateway-standalone/gravitee-apim-gateway-standalone-container/pom.xml:
```diff
@@ -32,7 +32,7 @@
     <properties>
         <jetty94.wiremock.version>9.4.44.v20210927</jetty94.wiremock.version>
-        <gravitee-connector-http.version>1.1.10</gravitee-connector-http.version>
+        <gravitee-connector-http.version>1.1.11</gravitee-connector-http.version>
```
in 3.18.x and upper, we should use the 2.0.x branch of connector-http.
See https://github.com/gravitee-io/gravitee-api-management/pull/2395

marcambier replied:
Nice catch, I cherry-picked both bump commits :+1:

Commit:
feat: improve authentication workflow of subscription based plans

Plan based authentication handlers have to use the SubscriptionService of the gateway api, instead of SubscriptionRepositoryWrappers which has been removed.

ApiKey plans based authentication handler can now retrieve subscription and check its validity.

Commit:
fix: make gateway apply request timeout properly

VertxReactorTimeoutHandler.route was never called so, never applying request timeouts.

Regression appeared with this commit : https://github.com/gravitee-io/gravitee-api-management/commit/5d29067211a8cb4f25bb85518e80a8ae0bcf7c12
Since this modification, the route method of the wrapped handler is called instead of the one handling timeouts.
